### DISCUSSION TITLE
使用 Cloudflare Workers 替代 Core.php 实现反代

### DISCUSSION THREAD

**Discussion by freejishu** [freejishu/StatusLive#28] [Ideas]
StatusLive项目中，由于 UptimeRobot API 对请求速率存在限制，所以在访问量比较大的时后端需要一定的缓存。大多数云函数平台并没有提供类似 Redis 的键值读取功能，也不支持在本身函数范围内读写数据，但 Cloudflare Workers 除外。

Cloudflare Workers 相比其他的云函数平台，更适合 StatusLive 作为反代，原因如下：
- KV 功能是其特色，类似 Redis，支持 Workers 对其进行读写，非常适合用来作为缓存；
- 免费账户也享有一定的 Workers 配额，足够中小型站点使用；
- UptimeRobot 的线路亦是 Cloudflare 线路，API 请求极快；
- 托管后的 SLA 由 Cloudflare 负责；
- 但已知的**缺点**：中国大陆地区的连接性可能不稳定

之前已有小伙伴提出此需求（ https://github.com/freejishu/StatusLive/discussions/15#discussioncomment-3489367 ），本篇指南将指导您如何使用 Cloudflare Workers 进行反代，代替 `core.php` 文件。

## 1、拥有一个 Cloudflare 账户

- 使用 Workers 之前，您需要拥有一个 Cloudflare 账户，可前往 Cloudflare 官网进行注册。

## 2、创建一个 Workers

- 登录控制台，找到左侧菜单 Workers → Overview → Create a Service：

![image](https://user-images.githubusercontent.com/11630435/207415073-0ee02b42-39bd-4237-98a1-5aad1260e20d.png)

- 键入您想要的前缀域名，并选择`HTTP handler/HTTP 处理程序`，然后 Create service：

![image](https://user-images.githubusercontent.com/11630435/207413637-87925723-0044-4013-82bc-66c8e9d4a3cc.png)

## 3、载入代码

> 除了使用 Cloudflare Web 页面进行编辑和部署，也可以使用 Cloudflare 提供的 [Wrangler CLI](https://dash.cloudflare.com/3a87e604034fb3787269411ae5f38e8c/workers/cli) 进行代码编写和部署，详情参照相关指南。

- 点击刚刚创建的 Workers，随后点击Quick Edit：

![image](https://user-images.githubusercontent.com/11630435/207415426-4395279d-540f-44a8-8d47-d6a20d2e9e07.png)

![image](https://user-images.githubusercontent.com/11630435/207415650-f0fd8e38-6a87-461e-83bc-8f27b3fce698.png)

- 在左侧的代码框中，粘贴如下代码：

```

const api_key = 'ur609264-e190f91e2cae9f70e3e9b201';//你的apikey

const cache_time = 10;//缓存时间（秒）

const url = 'https://api.uptimerobot.com/v2/getMonitors';


async function gatherResponse(response) {
    const { headers } = response;
    const contentType = headers.get('content-type') || '';
    if (contentType.includes('application/json')) {
        return JSON.stringify(await response.json());
    }
    return response.text();
}

async function readRequestBody(request) {
    const { headers } = request;
    const contentType = headers.get('content-type') || '';

    if (contentType.includes('application/json')) {
        return JSON.stringify(await request.json());
    } else if (contentType.includes('application/text')) {
        return request.text();
    } else if (contentType.includes('text/html')) {
        return request.text();
    } else if (contentType.includes('form')) {
        const formData = await request.formData();
        const body = {};
        for (const entry of formData.entries()) {
        body[entry[0]] = entry[1];
        }
        return JSON.stringify(body);
    } else {
        return 'a file';
    }
}

async function handleRequest(request) {
    const timestamp = (Date.parse(new Date()))/1000;
    const lasttime = await statuslive.get("statuslive_lasttime");
    let results = '';
    let cache_tag = '';
    if (timestamp - lasttime <= cache_time) {
        results = await statuslive.get("statuslive_json_cache");
        cache_tag = 'Cache Hit - Time:'+lasttime;
    }else{
        const post_data = JSON.parse(await readRequestBody(request));
        post_data.api_key = api_key;

        const init = {
            headers: {
                'content-type': 'application/json;charset=UTF-8',
            },
            method: 'POST',
            body: await JSON.stringify(post_data)
        };
        const response = await fetch(url, init);
        results = await JSON.parse(await gatherResponse(response));

        for (let index = 0; index < results.monitors.length; index++) {
            results.monitors[index].url = "";
            results.monitors[index].http_username = "";
            results.monitors[index].http_password = "";
            results.monitors[index].port = "";
        }

        results = await JSON.stringify(results);
        await statuslive.put("statuslive_json_cache", results, {expirationTtl: (cache_time < 60 ? 60 : cache_time)})
        await statuslive.put("statuslive_lasttime", timestamp, {expirationTtl: (cache_time < 60 ? 60 : cache_time)})
        cache_tag = 'Cache Miss - Time:'+timestamp;
    }

    let response_init = {
        headers: {
            'content-type': 'application/json;charset=UTF-8',
            'Access-Control-Allow-Origin': '*',
            'Access-Control-Allow-Methods': '*',
            'Access-Control-Allow-Credentials': 'true',
            'Access-Control-Allow-Headers': 'Content-Type',
            'Access-Control-Expose-Headers': '*',
            'StatusLive-Cache-Tag' : cache_tag
        }
    }
    return new Response(results, response_init);
}

addEventListener('fetch', event => {
    const request = event.request;
    const url = new URL(request.url);

    if (request.method === 'OPTIONS') {
        event.respondWith(
            new Response(null, {
                status: 200,
                statusText: 'OK',
                headers: {
                    'content-type': 'application/json;charset=UTF-8',
                    'Access-Control-Allow-Origin': '*',
                    'Access-Control-Allow-Methods': '*',
                    'Access-Control-Allow-Credentials': 'true',
                    'Access-Control-Allow-Headers': 'Content-Type',
                    'Access-Control-Expose-Headers': '*'
                }
            })
        );
    } else if (request.method === 'POST') {
        event.respondWith(handleRequest(request));
    } else {
        event.respondWith(
            new Response(null, {
                status: 405,
                statusText: 'Method Not Allowed',
            })
        );
    }

});

```

- 随后点击**Save and deploy**按钮保存并部署。

![image](https://user-images.githubusercontent.com/11630435/207518942-804580d6-d896-4104-b9cc-f8f9de5ee246.png)

## 4、新建并绑定 KV 

- 在 Dashboard 左侧菜单中找到 Workers → KV ，点击 Create namespace 按钮，在出现的文本框中输入你想要的 KV 命名空间，最后点击 Add 按钮添加；

![image](https://user-images.githubusercontent.com/11630435/207519485-a0bdc322-cdf3-4f75-8527-48ef53686b29.png)

- 回到稍早前创建的 Worker 界面，找到 Setting 选项卡 → Variables → KV Namespace Bindings → Add binding 按钮；

![image](https://user-images.githubusercontent.com/11630435/207519273-5a600503-40b9-4255-a1d7-c32a445d239c.png)

- Variable name 填写 `statuslive`，KV Namespace 选择刚刚创建的 KV 命名空间，随后点击 Save and deploy 按钮；

![image](https://user-images.githubusercontent.com/11630435/207520584-7496a532-f5ed-436b-a3ee-cce2c5f3bf7c.png)

## （可选）5、绑定自己的域名

- 可以在 Triggers 选项卡中添加自己的域名。注意：只能添加到本账户已经托管到 Cloudflare 了的域名。

![image](https://user-images.githubusercontent.com/11630435/207521579-d46df02f-0ed1-4e23-8f60-47f53fffd76b.png)

## 完成！

现在，将您的 Workers 域名填写到 `conf.json` 的 `config_proxy_link` 字段即可。

```
"config_proxy_link": "https://xxxxx.xxxxx.workers.dev/",  //填写Workers域名
```

或者，将其作为环境变量，进行 build & deploy。

**Comment by MoForgt** [comment]
代码有问题，不能部署到cloudflare中

**Reply by MoForgt** [reply]
因为里面调用了kv的存储名称，会有报错不能部署，得返回再重新进才行。。。

**Reply by freejishu** [reply]
理论上应该设置了KV之后，就可以一次性部署了，不如详细说说部署的步骤，看是否还有可以优化的地方？
